packages/internals/src/engine-commands/getEnginesMetaInfo.ts

```diff
@@ -0,0 +1,168 @@
+   * Resolve `resolveEngine` promises (that can never fail) and forward a reference to
+   * the engine name in the promise result.
+   */
+  const enginePromises = engines.map(({ name, type }) => {
+    return resolveEngine(type).then((result) => [name, result])
+  })
+  const engineMatrix: BinaryInfoMatrix = await Promise.all(enginePromises).then(Object.fromEntries)
+
+  const engineDataAcc = engines.map(({ name }) => {
+    const [engineInfo, errors] = getEnginesInfo(engineMatrix[name])
+    return [{ [name]: engineInfo } as { [name in keyof BinaryInfoMatrix]: string }, errors] as const
+  })
+
+  // map each engine to its version
+  const engineMetaInfo: {
+    'query-engine': string
+    'migration-engine': string
+    'introspection-engine': string
+    'format-binary': string
+  }[] = engineDataAcc.map((arr) => arr[0])
+
+  // keep track of any error that has occurred, if any
+  const enginesMetaInfoErrors: Error[] = engineDataAcc.flatMap((arr) => arr[1])
+
+  return [engineMetaInfo, enginesMetaInfoErrors] as const
+}
+
+export function getEnginesInfo(enginesInfo: EngineInfo): readonly [string, Error[]] {
+  // if the engine is not found, or the version cannot be retrieved, keep track of the resulting errors.
+  const errors = [] as Error[]
+
+  // compute message displayed when an engine is resolved via env vars
+  const resolved = match(enginesInfo)
+    .with({ fromEnvVar: P.when(O.isSome) }, (_engineInfo) => {
+      return `, resolved by ${_engineInfo.fromEnvVar.value}`
+    })
+    .otherwise(() => '')
+
+  // compute absolute path of an engine, returning an error message and populating
+  // `errors` if it fails
+  const absolutePath = match(enginesInfo)
+    .with({ path: P.when(E.isRight) }, (_engineInfo) => {
+      return _engineInfo.path.right
+    })
+    .with({ path: P.when(E.isLeft) }, (_engineInfo) => {
+      // the binary/library can't be found
+      errors.push(_engineInfo.path.left)
+      return 'E_CANNOT_RESOLVE_PATH' as const
+    })
+    .exhaustive()
+
+  // compute version (git hash) of an engine, returning an error message and populating
+  // `errors` if it fails
+  const version = match(enginesInfo)
+    .with({ version: P.when(E.isRight) }, (_engineInfo) => {
+      return _engineInfo.version.right
+    })
+    .with({ version: P.when(E.isLeft) }, (_engineInfo) => {
+      // the binary/library exists, but extracting the version failed
+      errors.push(_engineInfo.version.left)
+      return 'E_CANNOT_RESOLVE_VERSION_FROM_ENGINE' as const
+    })
+    .exhaustive()
+
+  const versionMessage = `${version} (at ${path.relative(process.cwd(), absolutePath)}${resolved})`
+  return [versionMessage, errors] as const
+}
+
+/**
+ * An engine path read from the environment is valid only if it exists on disk.
+ * @param pathFromEnv engine path read from process.env
+ */
+function isPathFromEnvValid(pathFromEnv: string | undefined): pathFromEnv is string {
+  return !!pathFromEnv && fs.existsSync(pathFromEnv)
+}
```

Comment:
Note: The following wouldn't infer types correctly (typescript prefers using `!!` to `Boolean(-)`):

```ts
return Boolean(pathFromEnv) && fs.existsSync(pathFromEnv)
```